### ISSUE TITLE
[Bug/Design] `SGLANG_SANITIZE_NAN_LOGITS`: a fully-NaN logits row becomes a uniform-random sample — garbage streamed to clients as normal output; proposal: opt-in per-request abort

### ISSUE THREAD

**Issue by jetd1** [sgl-project/sglang#33187]
## Summary

Following up on #32968 (transient NaN-contaminated logits on Kimi-K3): we enabled `SGLANG_SANITIZE_NAN_LOGITS=1` (#28829) in production to stop device-side assert crashes. It does stop the crashes — but it exposes a failure mode that is arguably **worse than the crash** for real users:

When an **entire logits row** is NaN, `nan_to_num_(nan=-1e30)` maps every lane to the same value. Softmax over a constant row is a **uniform distribution over the full vocabulary** (163,840 lanes for K3), so the sampler draws pure random tokens:

- The output is multilingual BPE salad (CJK/Latin/Cyrillic/code fragments mixed), streamed to the client as a **healthy 200 response** — no error, no signal, indistinguishable from real output at the serving layer.
- P(EOS) ≈ 1/163,840 per step, so generation essentially never terminates before `max_tokens`.
- The NaN'd step's KV is written to the request's KV cache, so **every subsequent step of that request is also corrupted** — the remainder of the response is garbage, matching what we observe.
- Reserved/special tokens appear as literal text (we observed `<|reserved_token_163672|>` in a delivered response — a token that is effectively unreachable under any healthy distribution; its presence is a reliable smoking gun for near-uniform sampling).

This hit a real interactive user session in our deployment. Sample of delivered output (thinking channel; surrounding user context omitted):

```
The049带点 accessingGEO-leading皮肤和怀抱萌萌 LarsMMDD Ptraky anthrop:"撕心观感ronesTester())))
getitem信息网初级阶段白开水捕捉到<Element不是要CDATApads_ACCESS ingred na_Act坐上温开水日后的有大
строй白萝卜—拍里边征兆在基层政策支持短裙.Hex弄虚rement fireplaces中央的angent…
```

## Environment

Same deployment as #32968: `kimi-k3` branch @ `f37baeb52f` + cherry-picks of #32477 and #33003, image lineage `lmsysorg/sglang:kimi-k3-c6ad1f26-20260729`, single-node TP=8 on 8× Blackwell-class GPUs (SM10x), four identical nodes. **Speculative decoding disabled** (relevant: these events are on the plain prefill→first-token-sampling path, independent of DSPARK).

## Observed events

Overnight after enabling the flag fleet-wide: **≥3 full-row NaN events across 3 of 4 nodes** (22:26Z, 23:07Z, 23:34Z), each logging

```
NaN detected in sampler: next_token_logits; values were sanitized before sampling. …
```

on **all 8 TP ranks simultaneously**. The events we correlated with decode logs coincide with a small prefill batch (e.g. a 64-token new sequence, `cached=0`) at first-token sampling. Zero asserts/crashes since enabling the flag (that part works as designed).

Note this count is a **lower bound**: `_AsyncNanWarner` is a process-lifetime one-shot latch ("Logged once; further occurrences are silent"), so a second event on the same worker is invisible. Operationally this also means the flag currently provides almost no telemetry.

## Why neither existing mode is adequate in production

| Mode | Blast radius | Client experience |
|---|---|---|
| default (assert) | whole engine: all in-flight requests die, ~20 min reload for a large model | connection dropped, no explanation |
| `SANITIZE_NAN_LOGITS=1` | zero crash | **garbage silently delivered as normal output**; poisoned KV can persist for the request; no signal to retry |

#28829 deliberately framed these as orthogonal crash/heal axes. What's missing is a third option with a *request-sized* blast radius.

Precedent for the phenotype: #30989 (`!!!!` collapse, NVFP4) and #22087 (garbled multilingual mixing, GDN kernel) — same visible symptom family, both ultimately kernel bugs. Root causes should of course be fixed in kernels; but transient NaN sources keep appearing (#32968 remains open), and the serving layer needs a sane containment mode in the meantime.

## Proposal: `SGLANG_ABORT_ON_NAN_LOGITS` (opt-in third mode)

Full-row NaN → **abort only the affected request(s)** with a retriable 5xx; partial-row NaN → sanitize as today (healthy lanes dominate after `-1e30`).

Sketch (validated against current `kimi-k3` branch code):

1. In `Sampler._preprocess_logits`, before `nan_to_num_`: compute `full_nan_rows = torch.isnan(logits).all(-1)` (gated on the env; zero overhead when off). Sanitize afterwards as today so downstream kernels never see NaN.
2. Carry the row mask to batch-result processing on the **existing** token-id device→host sync — no new GPU↔CPU syncs.
3. Scheduler: for flagged rows, `req.to_finish = FINISH_ABORT("Transient numerical error (NaN logits); please retry", 5xx)` — the same per-request machinery the running-timeout abort already uses.
4. **Cache hygiene** (the part that makes this more than a one-liner):
   - skip the radix-cache insert for the aborted request — otherwise the poisoned KV becomes **shareable via prefix matching** and a retry/prefix-sibling can read NaN KV;
   - correctly release hybrid state (Mamba/KDA) for the aborted request;
   - consider hierarchical-cache write-through interactions (host-cache copies of poisoned pages).
5. Per-event (throttled) logging with rid/batch context, replacing the one-shot latch for this path — this doubles as telemetry for tracking transient NaN sources like #32968.

We have an implementation of this in progress and intend to submit a PR. Filing this first for design feedback: flag naming/default, choice of status code (we lean 503 to match the running-timeout abort), interaction with the overlap scheduler and speculative decoding, and whether partial-row events deserve a counter/log line as well.
